### PR TITLE
[SPARK-37228][SQL][PYTHON] Implement DataFrame.mapInArrow in Python

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to implement `DataFrame.mapInArrow` that allows users to apply a function with PyArrow record batches such as:

```python
def do_something(iterator):
    for arrow_batch in iterator:
        # do something with `pyarrow.RecordBatch` and create new `pyarrow.RecordBatch`.
        # ...
        yield arrow_batch

df.mapInArrow(do_something, df.schema).show()
```

The general idea is simple. It shares the same codebase of `DataFrame.mapInPandas` except the pandas conversion logic.

This PR also piggy-backs:
- Removes the check in `spark.udf.register` on `SQL_MAP_PANDAS_ITER_UDF`. This type is only used for `DataFrame.mapInPandas` internally, and it cannot be registered as a SQL UDF
- Removes the type hints for `pandas_udf` that is used for internal purposes such as `SQL_MAP_PANDAS_ITER_UDF` and `SQL_COGROUPED_MAP_PANDAS_UDF`. Both cannot be used for `pandas_udf` as a SQL expression and it should be hidden to end users.

Note that documentation will be done in another PR.

### Why are the changes needed?

For usability and technical problems. Both are elabourated in more details at SPARK-37227.
Please also see the discussions at https://github.com/apache/spark/pull/26783.

### Does this PR introduce _any_ user-facing change?

Yes, this PR adds a new API:

```python
import pyarrow as pa

df = spark.createDataFrame(
    [(1, "foo"), (2, None), (3, "bar"), (4, "bar")], "a int, b string")

def func(iterator):
    for batch in iterator:
        # `batch` is pyarrow.RecordBatch.
        yield batch

df.mapInArrow(func, df.schema).collect()
```

### How was this patch tested?

Manually tested, and unit tests were added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
This PR is ready for a review.

**Comment by HyukjinKwon** [comment]
cc @zero323, @ueshin, @BryanCutler, @viirya, @jpivarski, @revans2, @tgravescs, @LucaCanali, @Tagar

**Comment by Tagar** [comment]
cc @lgray

**Comment by lgray** [comment]
Thanks for the CC - I'll have to try this out soon, but will require updating our stuff to spark 3 (should be easy). It is, however, not the highest priority.

**Comment by jpivarski** [comment]
**Thank you!!!** This is exactly what we've been looking for; with this in place, we'll be able to process data like

```python
>>> import awkward as ak
>>> df = spark.createDataFrame([([0, 1, 2],), ([],), ([3, 4],), ([5],), ([6, 7, 8, 9],)], ("x",))
>>> def func(iterator):
...     for batch in iterator:
...         akarray = ak.from_arrow(batch)
...         combinations = ak.combinations(akarray["x"], 2, axis=1)
...         yield from ak.to_arrow_table(ak.Array({"comb": combinations})).to_batches()
... 
>>> df.mapInArrow(func, df.schema)
```

(Assuming the data with nested structure, such as the `DataFrame[x: array<bigint>]` above, are created in some scalable way, most likely by [Laurelin](https://github.com/spark-root/laurelin). Our ROOT data are full of these nested lists. [ak.combinations](https://awkward-array.readthedocs.io/en/latest/_auto/ak.combinations.html), etc. are Awkward Array functions for dealing with these things without Python iterations. Doing the same thing through Pandas turns all of our vectorized arrays into Python lists, but the Spark → Arrow → Awkward → Arrow → Spark transformation keeps them as contiguous arrays with minimal memory and CPU footprint...)

**Comment by lgray** [comment]
Thanks for checking this @jpivarski !!! Looks nice!

**Comment by PerilousApricot** [comment]
Hi @HyukjinKwon - thanks for this PR! I had a question about the python-facing API. Is it possible to have something similar to how current python/pandas UDFs' withColumn() can tell spark which columns are necessary for the inputs and pass the remainder of the columns through to the resulting DF without round-tripping them through python? For example (open-coding this, might be missing some syntax):

```python

df = spark.createDataFrame(
    [(1, "foo"), (2, None), (3, "bar"), (4, "bar")], "a int, b string")

@udf(IntType())
def squared(x):
    return x * x

df = df.withColumn("squared", squared["a"]))
```

Would only ser/de the "a" column through the function, but the remaining column "b" would be in the resulting dataframe.

**Comment by HyukjinKwon** [comment]
Sounds like that's close to the original proposal of  #26783. I was skeptical about that idea and it might not be worthwhile because:

- We will have to take care of other restrictions, and variants like aggregation, window, etc all together, and thought that might not be worthwhile
- I still think that Arrow is rather an internal format instead of user-facing.

So these made me propose one (developer) API that doesn’t require considering other restrictions
(e.g., the length of input should be the same as output's in case of scalar UDF in `select`) or variants.
In addition, I am pretty sure the current API can do what UDFs could do although it might be less optimizing.

However, I am open to other suggestions.

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/sql/pandas/map_ops.py [hunk 1]
```diff
@@ -91,6 +91,68 @@ def mapInPandas(
         jdf = self._jdf.mapInPandas(udf_column._jc.expr())  # type: ignore[operator]
         return DataFrame(jdf, self.sql_ctx)
 
+    def mapInArrow(
+        self, func: "ArrowMapIterFunction", schema: Union[StructType, str]
+    ) -> "DataFrame":
+        """
+        Maps an iterator of batches in the current :class:`DataFrame` using a Python native
+        function that takes and outputs a PyArrow's `RecordBatch`, and returns the result as a
+        :class:`DataFrame`.
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/map_ops.py:100]: result as a `RecordBatch` ?

**Review by BryanCutler — APPROVED** [review]
LGTM, this is great @HyukjinKwon thanks for working on it!

File: python/pyspark/sql/pandas/map_ops.py (hunk 1 above)
Comment (BryanCutler) [reply python/pyspark/sql/pandas/map_ops.py:100]: Oh I think I read it wrong, result is a PySpark `DataFrame`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/MapInBatchExec.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * A relation produced by applying a function that takes an iterator of batches
+ * such as pandas DataFrame or PyArrow's record batches, and outputs an iterator of them.
+ */
+trait MapInBatchExec extends UnaryExecNode {
+  protected val func: Expression
+  protected val pythonEvalType: Int
+
+  private val pandasFunction = func.asInstanceOf[PythonUDF].func
```
Comment (galv) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/MapInBatchExec.scala:41]: This seems like an incorrect name, given that pandas is not used for execution here.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/MapInBatchExec.scala:41]: thanks for pointing this out.

File: python/pyspark/worker.py [hunk 3]
```diff
@@ -364,12 +373,15 @@ def read_udfs(pickleSer, infile, eval_type):
...
     is_scalar_iter = eval_type == PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF
     is_map_iter = eval_type == PythonEvalType.SQL_MAP_PANDAS_ITER_UDF
+    is_map_arrow_iter = eval_type == PythonEvalType.SQL_MAP_ARROW_ITER_UDF
 
-    if is_scalar_iter or is_map_iter:
+    if is_scalar_iter or is_map_iter or is_map_arrow_iter:
         if is_scalar_iter:
             assert num_udfs == 1, "One SCALAR_ITER UDF expected here."
         if is_map_iter:
             assert num_udfs == 1, "One MAP_ITER UDF expected here."
```
Comment (viirya) [inline python/pyspark/worker.py:382]: Should it be `MAP_PANDAS_ITER` instead of `MAP_ITER`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonMapInArrowExec.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.spark.sql.execution.SparkPlan
+
+/**
+ * A relation produced by applying a function that takes an iterator of PyArrow's record batches
+ * and outputs an iterator of PyArrow's record batches.
+ *
+ * This is somewhat similar with [[FlatMapGroupsInPandasExec]] and
+ * `org.apache.spark.sql.catalyst.plans.logical.MapPartitionsInRWithArrow`
+ */
+case class PythonMapInArrowExec(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonMapInArrowExec.scala:31]: `MapInArrowExec`?

**Review by viirya — COMMENTED** [review]
This looks good. I have a few questions mostly about the naming.

**Comment by tgravescs** [comment]
thanks @HyukjinKwon, I think this API makes sense for our use cases at the moment

File: python/pyspark/sql/udf.py [hunk 5]
```diff
@@ -405,12 +409,10 @@ def register(
                 PythonEvalType.SQL_SCALAR_PANDAS_UDF,
                 PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF,
                 PythonEvalType.SQL_GROUPED_AGG_PANDAS_UDF,
-                PythonEvalType.SQL_MAP_PANDAS_ITER_UDF,
             ]:
                 raise ValueError(
                     "Invalid f: f must be SQL_BATCHED_UDF, SQL_SCALAR_PANDAS_UDF, "
-                    "SQL_SCALAR_PANDAS_ITER_UDF, SQL_GROUPED_AGG_PANDAS_UDF or "
-                    "SQL_MAP_PANDAS_ITER_UDF."
```
Comment (holdenk) [inline python/pyspark/sql/udf.py:413]: This change seems unrelated and a regression, did PANDAS_ITER_UDF not work here?

**Review by holdenk — COMMENTED** [review]
Thanks for working on this, dug through python code some questions will dig through scala side later. I think it's a useful feature especially  as more libraries accept arrow inputs.

**Comment by galv** [comment]
There's something this PR has made me begin to ponder. The "stream" object used by is created by the socket library's makefile() API: https://docs.python.org/3/library/socket.html#socket.socket.makefile

This means it is not a traditional file (i.e., the BSD socket API does not support posix read and write, so this is just a convenience provided by python). If a pipe were to be used instead of a socket, it seems conceivable that arrow data structures could be written to the pipe via the vmsplice() syscall, which would effectively do zero-copy movement of data from python to the JVM executor (I believe the virtual memory pages simply get assigned to the pipe file descriptor inside the kernel). My understanding was that the python worker.py process is always on the same machine as the JVM executor, so this seems like a reasonable speedup to consider.

**Comment by Tagar** [comment]
As an option if it makes sense - Arrow also comes with Plasma https://arrow.apache.org/blog/2017/08/08/plasma-in-memory-object-store/ 

> Plasma, an in-memory object store that is being developed as part of Apache Arrow 

Ray uses Plasma heavily. 
Would something like Plasma add scalability here for zero-copy shared-memory exchanges between jvm and python?

**Comment by HyukjinKwon** [comment]
> The "stream" object used by is created by the socket library's makefile() API: https://docs.python.org/3/library/socket.html#socket.socket.makefile

Sounds like an idea but I would like to avoid this approach for now because:
1. I am actually investigating a way of _true_ zero copy by shared memory between JVM and Python side. Using socket isn't actually true zero copy. It does copy from JVM to Python side although they are in a streaming approach.
2. Using socket is I think too low-level API, and current API should be able to allow users to use byte stream way within the UDF itself - the cost of wrapping Arrow instance on the top of the actual Arrow bytes would be trivial.
3.I think using higher level is more common than the socket approach.

**Comment by galv** [comment]
@Tagar it is conceivable to use something like plasma for zero-copy sharing of data between a spark JVM executor and the python worker. However, there are many hurdles that I see:

- You would need to implement a brand new allocator (in addition to the "Unsafe"/off-heap allocator and the "JVM native memory" allocator), which creates a Memory-mapped file, and the does all allocations within that file.
- Shared memory is not a resource which is scheduled by most schedulers, AFAIK. I know yarn doesn't handle it. I also know that the spark standalone scheduler doesn't handle it. Maybe k8s does, but I don't know. As an alternative to using shared memory, the JVM executor can create a memory-mapped file (as above), but instead of making it in shared memory, it can make it in its own private memory space, and then *share* that file descriptor with its child worker.py process via memfd_create(): https://man7.org/linux/man-pages/man2/memfd_create.2.html . However, memfd_create() is a relatively now syscall which is only in linux! (That is probably okay if this is behind a configuration option.)
- Spark itself would have to use Arrow's format as its execution format to do zero-copying sharing between python and java. Unfortunately, almost all spark SQL operations are done on its InternalRow format (except for some special cases involving parquet and other columnar file stores I think), which is not a columnar format. I am almost certain that @tgravescs and others have thought about this before and made JIRA issues about supporting a columnar format (possibly arrow, possibly not) in Spark, but I don't know what happened to it.

All that is meant to say that true zero-copy seemingly requires touching a lot of parts of Spark. My proposal to move from using a BSD socket to using a pipe and using vmsplice just (potentially) fixes one thing: the copy done from python to JVM. It cannot even speed up the copying from the pipe to python's memory because of limitations of the vmsplice() syscall.

> I am actually investigating a way of true zero copy by shared memory between JVM and Python side. Using socket isn't actually true zero copy. It does copy from JVM to Python side although they are in a streaming approach.

@HyukjinKwon I'm afraid there has been a misunderstanding. I agree that using a socket isn't zero copy at all. I am just saying that it is what it is done *today*. Anyway, I would be very interested in learning more about your thoughts on zero copy between JVM and python.

**Comment by HyukjinKwon** [comment]
Let me address the rest of comments on codes by EOD today.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonMapInArrowExec.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonMapInArrowExec.scala:31]: Ah, I will implement `MapInArrowExec` in Scala side at SPARK-37229. Both will have a different implementation that cannot share the common codes so let me keep it as is like `PythonForeachWriter` vs `ForeachWriter`

File: python/pyspark/sql/udf.py (hunk 5 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py:413]: Oh I had to clarify this. Yeah `SQL_MAP_PANDAS_ITER_UDF` is only used for `DataFrame.mapInPandas` which cannot be registered as a UDF. I will clarify this in Pr description.

File: python/pyspark/sql/pandas/functions.pyi [hunk 6]
```diff
@@ -130,39 +124,3 @@ def pandas_udf(
 def pandas_udf(
     f: Union[AtomicDataTypeOrString, ArrayType], *, functionType: PandasGroupedAggUDFType
 ) -> Callable[[PandasGroupedAggFunction], UserDefinedFunctionLike]: ...
-@overload
-def pandas_udf(
-    f: PandasMapIterFunction,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/functions.pyi:135]: Removed this because these type hints are not supposed to be exposed to end users (see also PR description)

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Will merge this one in few more days if there aren't more comments.

**Review by dongjoon-hyun — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Thanks guys!
